### ISSUE TITLE
clippy::unnecessary_cast does not warn on basic usage

### ISSUE THREAD

**Issue by frewsxcv** [rust-lang/rust-clippy#9562]
### Summary

When casting the result of a function to the same type, `clippy::unnecessary_cast` doesn't warn

### Lint Name

clippy::unnecessary_cast

### Reproducer

I tried this code:

```rust
fn foo() -> f32 { 0. }

fn main() {
    let num = foo() as f32;
}
```

I expected to see this happen: Warning that the cast is unnecessary

Instead, this happened: No warning

**Comment by xFrednet** [comment]
Issue actions are done with `@rustbot`. I can assign you directly :)
